src/trees/FunctionTree.cpp

```diff
@@ -657,11 +658,26 @@ template <int D> void FunctionTree<D>::appendTreeNoCoeff(MWTree<D> &inTree) {
         MWNode<D> *inNode = instack.back();
         instack.pop_back();
         if (inNode->getNChildren() > 0) {
+            thisNode->clearIsEndNode();
             if (thisNode->getNChildren() < inNode->getNChildren()) thisNode->createChildren(false);
             for (int i = 0; i < inNode->getNChildren(); i++) {
                 instack.push_back(inNode->children[i]);
                 thisstack.push_back(thisNode->children[i]);
             }
+        } else {
+            // construct EndNodeTable for "This", starting from this branch
+            // This could be done more efficiently, if it proves to be time consuming
+            std::vector<MWNode<D> *> branchstack;   // local stack starting from this branch
+            branchstack.push_back(thisNode);
+            while (branchstack.size() > 0) {
+                MWNode<D> *branchNode = branchstack.back();
+                branchstack.pop_back();
+                if (branchNode->getNChildren() > 0) {
+                    for (int i = 0; i < branchNode->getNChildren(); i++) {
+                        branchstack.push_back(branchNode->children[i]);
+                    }
+                } else this->endNodeTable.push_back(branchNode);
+            }
```

Comment:
There is a `resetEndNodeTable()` function in MWTree which you could call after the tree is fully constructed., assuming the endNodeTable does not have to up to date _during_ this process. It will be less efficient, though, as it will traverse the tree one more time.